agent/options.go

```diff
@@ -0,0 +1,195 @@
+//  Copyright (c) 2018 Cisco and/or its affiliates.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at:
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package agent
+
+import (
+	"errors"
+	"reflect"
+	"time"
+
+	"github.com/ligato/cn-infra/core"
+	"github.com/ligato/cn-infra/logging/logrus"
+)
+
+// Options specifies the Version, MaxStartupTime, and Plugin list for the Agent
+type Options struct {
+	Version        string
+	MaxStartupTime time.Duration
+
+	Plugins []core.PluginNamed
+}
+
+func newOptions(opts ...Option) Options {
+	opt := Options{
+		Version:        "dev",
+		MaxStartupTime: time.Second * 15,
+	}
+
+	for _, o := range opts {
+		o(&opt)
+	}
+
+	return opt
+}
+
+// Option is a function that operates on an Agent's Option
+type Option func(*Options)
+
+// MaxStartupTime returns an Option that sets the MaxStartuptime option of the Agent
+func MaxStartupTime(d time.Duration) Option {
+	return func(o *Options) {
+		o.MaxStartupTime = d
+	}
+}
+
+// Version returns an Option that sets the version of the Agent to the string v
+func Version(v string) Option {
+	return func(o *Options) {
+		o.Version = v
+	}
+}
+
+// Plugins creates an Option that adds a list of Plugins to the Agent's Plugin list
+func Plugins(plugins ...core.PluginNamed) Option {
+	return func(o *Options) {
+		o.Plugins = append(o.Plugins, plugins...)
+	}
+}
+
+// DescendantPlugins creates an Option that adds all of the Descendants of a Plugin Recursively to the Agent's
+// Plugin list
+func DescendantPlugins(plugin core.Plugin) Option {
```

Comment:
Hmm, perhaps `AllPlugins` might be simpler? And would actually mean that it will init all the plugins nested in this one.

Reply:
I can get behind AllPlugins 👍